### PR TITLE
Improve error message for name/label validation.

### LINKED ISSUE #37654: PersistentVolume metadata.name is required to be lower case
**Kubernetes version** (use `kubectl version`):

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.6", GitCommit:"ae4550cc9c89a593bcda6678df201db1b208133b", GitTreeState:"clean", BuildDate:"2016-08-26T18:13:23Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:10:32Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- AWS

```
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

```
$ uname -a
Linux ip-10-8-1-99 3.13.0-91-generic #138-Ubuntu SMP Fri Jun 24 17:00:34 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

- cluster bootstrapped with Kops

**What happened**:

I tried to create a PersistentVolume resource and I received the following error output

```
$ cat es-efs-volume.yml 
apiVersion: v1
kind: PersistentVolume
metadata:
  name: Nfs
spec:
  capacity:
    storage: 50Gi
  accessModes:
    - ReadWriteMany
  nfs:
    server: us-west-2a.fs-64738265.efs.us-west-2.amazonaws.com
    path: "/"

$ kubectl create -f es-efs-volume.yml
The PersistentVolume "Nfs" is invalid.
metadata.name: Invalid value: "Nfs": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)* (e.g. 'example.com')
```

**What you expected to happen**:

I am given a newly created PersistentVolume resource

**How to reproduce it** (as minimally and precisely as possible):

Create the resource with capital letters as the value for metadata.name

**Anything else do we need to know**:

The fact that this key requires all lower case characters in the name is not indicated by the error output. I only understood this was the error after reading through the whole regex. I am of the opinion that reading regexes to understand error output is not a nice task to expect all humans to perform. Also, why is the name being validated through a regex?

### PR DESCRIPTION
**Pull request by xingzhou**
Instead of just providing regex in name/label validation error output, we need to add the naming rules of the name/label, which is more end-user readable.

Fixed #37654

### CODE REVIEW

**Comment by xingzhou** [comment]
have rebased the code, @smarterclayton, would you please help review the patch? thanks!

File: pkg/util/validation/validation.go [hunk 1]
```diff
@@ -80,12 +82,13 @@ func IsValidLabelValue(value string) []string {
...
 	}
 	if !labelValueRegexp.MatchString(value) {
-		errs = append(errs, RegexError(labelValueFmt, "MyValue", "my_value", "12345"))
+		errs = append(errs, RegexError(labelValueErrMsg, labelValueFmt, "MyValue", "my_value", "12345"))
 	}
 	return errs
 }
 
 const dns1123LabelFmt string = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"
+const dns1123LabelErrMsg string = "a valid DNS (RFC 1123) label must consist of lower case alphanumeric character or '-', and must start/end with an alphanumeric character"
```
Comment (smarterclayton) [inline pkg/util/validation/validation.go]: Pluralize the first `characters`: `must consist of lower case alphanumeric characters or ...`

File: pkg/util/validation/validation.go [hunk 2]
```diff
@@ -98,12 +101,13 @@ func IsDNS1123Label(value string) []string {
...
 	}
 	if !dns1123LabelRegexp.MatchString(value) {
-		errs = append(errs, RegexError(dns1123LabelFmt, "my-name", "123-abc"))
+		errs = append(errs, RegexError(dns1123LabelErrMsg, dns1123LabelFmt, "my-name", "123-abc"))
 	}
 	return errs
 }
 
 const dns1123SubdomainFmt string = dns1123LabelFmt + "(\\." + dns1123LabelFmt + ")*"
+const dns1123SubdomainErrorMsg string = "a valid DNS (RFC 1123) subdomain label must consist of lower case alphanumeric character, '-' or '.', and must start/end with an alphanumeric character"
```
Comment (smarterclayton) [inline pkg/util/validation/validation.go]: Pluralize first `characters`: `must consist of lower case alphanumeric characters, ...`

File: pkg/util/validation/validation.go [hunk 3]
```diff
@@ -144,6 +149,7 @@ func IsDNS1035Label(value string) []string {
 // - valid: *.bar.com, *.foo.bar.com
 // - invalid: *.*.bar.com, *.foo.*.com, *bar.com, f*.bar.com, *
 const wildcardDNS1123SubdomainFmt = "\\*\\." + dns1123SubdomainFmt
+const wildcardDNS1123SubdomainErrMsg = "a valid wildcard DNS (RFC 1123) subdomain label must start with '*.', followed by a valid DNS subdomain name, which must consist of lower case alphanumeric character, '-' or '.' and end with an alphanumeric character"
```
Comment (smarterclayton) [inline pkg/util/validation/validation.go]: Pluralize first `character`: `must consist of lower case alphanumeric characters, ...`

File: pkg/util/validation/validation.go [hunk 4]
```diff
@@ -155,20 +161,21 @@ func IsWildcardDNS1123Subdomain(value string) []string {
...
 	}
 	if !wildcardDNS1123SubdomainRegexp.MatchString(value) {
-		errs = append(errs, RegexError(wildcardDNS1123SubdomainFmt, "*.example.com"))
+		errs = append(errs, RegexError(wildcardDNS1123SubdomainErrMsg, wildcardDNS1123SubdomainFmt, "*.example.com"))
 	}
 	return errs
 }
 
 const cIdentifierFmt string = "[A-Za-z_][A-Za-z0-9_]*"
+const identifierErrMsg string = "a valid identifier must start with alphabetic character or '_', followed by a string consist of alphanumeric character or '_'"
```
Comment (smarterclayton) [inline pkg/util/validation/validation.go]: `followed by a string consisting of alphanumeric characters ...`

**Comment by smarterclayton** [comment]
@k8s-bot ok to test

**Comment by xingzhou** [comment]
Thanks @smarterclayton for your review, has updated the code according to your comments, please take a look.

**Comment by xingzhou** [comment]
@k8s-bot unit test this

**Comment by xingzhou** [comment]
@smarterclayton, not sure we need to re-test this, but I did not see the failed test case is relevant to the changes in the patch.

File: pkg/util/validation/validation.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ import (
 const qnameCharFmt string = "[A-Za-z0-9]"
 const qnameExtCharFmt string = "[-A-Za-z0-9_.]"
 const qualifiedNameFmt string = "(" + qnameCharFmt + qnameExtCharFmt + "*)?" + qnameCharFmt
+const qualifiedNameErrMsg string = "must consist of alphanumeric characters, '-', '_' or '.', and must start/end with an alphanumeric character"
```
Comment (thockin) [inline pkg/util/validation/validation.go]: start and end"

**Comment by xingzhou** [comment]
@smarterclayton and @thockin, have fixed the code according to your comments, please take a look, thanks!

File: pkg/util/validation/validation.go [hunk 6]
```diff
@@ -61,12 +62,13 @@ func IsQualifiedName(value string) []string {
...
 	}
 	if !qualifiedNameRegexp.MatchString(name) {
-		errs = append(errs, "name part "+RegexError(qualifiedNameFmt, "MyName", "my.name", "123-abc"))
+		errs = append(errs, "name part "+RegexError(qualifiedNameErrMsg, qualifiedNameFmt, "MyName", "my.name", "123-abc"))
 	}
 	return errs
 }
 
 const labelValueFmt string = "(" + qualifiedNameFmt + ")?"
+const labelValueErrMsg string = "a valid label must be empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character"
```
Comment (thockin) [inline pkg/util/validation/validation.go]: "an empty string" or "the empty string"

File: pkg/util/validation/validation.go [hunk 7]
```diff
@@ -144,6 +149,7 @@ func IsDNS1035Label(value string) []string {
 // - valid: *.bar.com, *.foo.bar.com
 // - invalid: *.*.bar.com, *.foo.*.com, *bar.com, f*.bar.com, *
 const wildcardDNS1123SubdomainFmt = "\\*\\." + dns1123SubdomainFmt
+const wildcardDNS1123SubdomainErrMsg = "a valid wildcard DNS (RFC 1123) subdomain label must start with '*.', followed by a valid DNS subdomain name, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character"
```
Comment (thockin) [inline pkg/util/validation/validation.go]: "subdomain label" again

Comment (thockin) [inline pkg/util/validation/validation.go]: also "subdomain name" -> "subdomain"

File: pkg/util/validation/validation.go [hunk 8]
```diff
@@ -155,20 +161,21 @@ func IsWildcardDNS1123Subdomain(value string) []string {
...
 	}
 	if !wildcardDNS1123SubdomainRegexp.MatchString(value) {
-		errs = append(errs, RegexError(wildcardDNS1123SubdomainFmt, "*.example.com"))
+		errs = append(errs, RegexError(wildcardDNS1123SubdomainErrMsg, wildcardDNS1123SubdomainFmt, "*.example.com"))
 	}
 	return errs
 }
 
 const cIdentifierFmt string = "[A-Za-z_][A-Za-z0-9_]*"
+const identifierErrMsg string = "a valid identifier must start with alphabetic character or '_', followed by a string consist of alphanumeric characters or '_'"
```
Comment (thockin) [inline pkg/util/validation/validation.go]: "C identifier"

Comment (thockin) [inline pkg/util/validation/validation.go]: s/consist//

**Review by thockin — COMMENTED** [review]
sorry, a few nits still

**Comment by xingzhou** [comment]
@thockin, thanks for the comments. Have updated the code and tests based on your comments, please take a look, thanks

**Comment by xingzhou** [comment]
@thockin  and @smarterclayton, I did not see the relevance between the failed GKE smoke e2e tests and this patch, I've fixed the code based on your comments, would you please help re-kick the test? thanks

**Comment by thockin** [comment]
@k8s-bot gci gce e2e test this

File: pkg/util/validation/validation.go (hunk 1 above)
Comment (pires) [inline pkg/util/validation/validation.go]: DNS is case-insensitive so why is this here @xingzhou?

File: pkg/util/validation/validation.go (hunk 2 above)
Comment (pires) [inline pkg/util/validation/validation.go]: DNS is case-insensitive so why is this here @xingzhou?
